x/superfluid/keeper/migrate_test.go

```diff
@@ -0,0 +1,249 @@
+package keeper_test
+
+import (
+	"time"
+
+	"github.com/cosmos/cosmos-sdk/simapp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	stakingtypes "github.com/cosmos/cosmos-sdk/x/staking/types"
+
+	cl "github.com/osmosis-labs/osmosis/v14/x/concentrated-liquidity"
+
+	"github.com/osmosis-labs/osmosis/osmomath"
+	"github.com/osmosis-labs/osmosis/v14/x/gamm/pool-models/balancer"
+	gammtypes "github.com/osmosis-labs/osmosis/v14/x/gamm/types"
+	"github.com/osmosis-labs/osmosis/v14/x/superfluid/keeper"
+	"github.com/osmosis-labs/osmosis/v14/x/superfluid/types"
+)
+
+// We test migrating in the following circumstances:
+// 1. Migrating lock that is not superfluid delegated, not unlocking.
+// 2. Migrating lock that is not superfluid delegated, unlocking.
+// 3. Migrating lock that is superfluid delegated, not unlocking.
+// 4. Migrating lock that is superfluid undelegating, not unlocking.
+// 5. Migrating lock that is superfluid undelegating, unlocking.
+func (suite *KeeperTestSuite) TestUnlockAndMigrate() {
+	testCases := []struct {
+		name                     string
+		superfluidDelegated      bool
+		superfluidUndelegating   bool
+		unlocking                bool
+		percentOfSharesToMigrate sdk.Dec
+	}{
+		{
+			"lock that is not superfluid delegated, not unlocking",
+			false,
+			false,
+			false,
+			sdk.MustNewDecFromStr("0.9"),
+		},
+		{
+			"lock that is not superfluid delegated, unlocking",
+			false,
+			false,
+			true,
+			sdk.MustNewDecFromStr("0.6"),
+		},
+		{
+			"lock that is superfluid delegated, not unlocking",
+			true,
+			false,
+			false,
+			sdk.MustNewDecFromStr("1"),
+		},
+		{
+			"lock that is superfluid undelegating, not unlocking",
+			true,
+			true,
+			false,
+			sdk.MustNewDecFromStr("0.5"),
+		},
+		{
+			"lock that is superfluid undelegating, unlocking",
```

Comment:
Is it possible to test the "where it left" off part somehow>?

Reply:
Added test! https://github.com/osmosis-labs/osmosis/pull/4129/commits/04518037f42656afd74f111b82f9bdbfe96b541b